apps/hpc-ftsadmin/src/app/utils/fn-validations.ts

```diff
@@ -0,0 +1,375 @@
+  const locationNames = locations.map((loc) => loc.name).join(', ');
+
+  const hasValidLocation = formLocations.some((loc) =>
+    locationIds.has(valueToInteger(loc.value))
+  );
+  if (!hasValidLocation) {
+    return [
+      t.t(lang, (s) => s.components.flowForm.submitValidation.planLocation, {
+        entity: t.t(
+          lang,
+          (s) => s.components.flowsFilter.filters[`${direction}Locations`]
+        ),
+        plan: planVersion.name,
+        locations: locationNames,
+      }),
+    ];
+  }
+  return [];
+};
+const validatePlanLocation = async (
+  values: FlowFormType,
+  env: Environment,
+  lang: LanguageKey
+): Promise<string[]> => {
+  if (!values.fundingDestinationPlan && !values.fundingSourcePlan) {
+    return [];
+  }
+  return [
+    ...(await validatePlan(
+      values.fundingSourcePlan,
+      values.fundingSourceLocations,
+      lang,
+      env,
+      'source'
+    )),
+    ...(await validatePlan(
+      values.fundingDestinationPlan,
+      values.fundingDestinationLocations,
+      lang,
+      env,
+      'destination'
+    )),
+  ];
+};
+
+const validateFlowForWarnings = async (
+  values: FlowFormType,
+  env: Environment,
+  lang: LanguageKey
+) => {
+  const reportingDetailWarning = validateReportingDetails(values, lang);
+  if (reportingDetailWarning) {
+    toast.error(reportingDetailWarning, TOAST_CONFIG_ERROR);
+    return false;
+  }
+  const planLocationValidation = await validatePlanLocation(values, env, lang);
+  if (planLocationValidation.length) {
+    for (const validation of planLocationValidation) {
+      toast.error(validation, TOAST_CONFIG_ERROR);
+    }
+    return false;
+  }
+
+  const warnings: string[] = [
+    ...validateEarmarking(values, lang),
+    ...validateReportingOrganization(values, lang),
+    ...validateParentFlowAmountUSD(values, lang),
+    ...(await validateEmergency(values, env, lang)),
+  ];
+
+  for (const warning of warnings) {
+    if (!globalThis.confirm(warning)) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+export const validateFlowIsUnlinked = (flow: FlowFormType) => {
+  if (!flow.parentFlow && flow.childFlows.length === 0) {
+    return true;
+  }
+  return false;
+};
+
+export const validateFlow = async ({
+  values,
+  lang,
+  env,
+  pendingValuesHandled,
+  pendingValues,
+  isPending,
+}: {
+  values: FlowFormTypeValidated;
+  lang: LanguageKey;
+  env: Environment;
+  pendingValuesHandled: number;
+  pendingValues?: Partial<FlowFormType> | null;
+  isPending?: boolean;
+}) => {
+  const {
+    amountOriginalCurrency,
+    currency,
+    exchangeRate,
+    parentFlow,
+    childFlows,
+  } = values;
+
+  const originalCurrencyFields = [
+    ['amountOriginalCurrency', amountOriginalCurrency],
+    ['currency', currency],
+    ['exchangeRate', exchangeRate],
+  ] as const satisfies Array<
+    [
+      keyof FlowFormTypeValidated,
+      FlowFormTypeValidated[keyof FlowFormTypeValidated],
+    ]
```

Comment:
If we define type
```ts
type CurrencyFields = Extract<
  keyof FlowFormTypeValidated,
  'amountOriginalCurrency' | 'currency' | 'exchangeRate'
>;
```
this can become:
```suggestion
    [CurrencyFields, FlowFormTypeValidated[CurrencyFields]]
```